feat: add total records and bytes info in state message in dest cdk + impl dummy stats emitter

subodh1810 opened the pull request:
Issue : https://github.com/airbytehq/airbyte-internal-issues/issues/13096 and https://github.com/airbytehq/airbyte-internal-issues/issues/13095

subodh1810 committed:
feat: add total records and bytes info in state message in dest cdk

subodh1810 committed:
fix test

johnny-schmidt commented on airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/state/PipelineEventBookkeepingRouter.kt:
```diff
@@ -162,11 +166,15 @@ class PipelineEventBookkeepingRouter(
                         val streamWithKeyAndCount =
                             catalog.streams.map { stream ->
                                 val manager = syncManager.getStreamManager(stream.descriptor)
+                                totalRecords = totalRecords.plus(manager.readCount())
```
What you want to use is `CheckpointValue`, which currently conveys the committed record count per checkpoint and has a commented TODO for adding byte count.

Threading it through will be a little involved, let me gather the details...

johnny-schmidt replied:
4.  That class writes the checkpoint ids to a queue (in `handleOutput`), which is read by `UpdateBatchStateTask`, which in turn updates the StreamManager. The changes along that path should be obvious.

johnny-schmidt commented on airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/state/PipelineEventBookkeepingRouter.kt:
```diff
@@ -162,11 +166,15 @@ class PipelineEventBookkeepingRouter(
                         val streamWithKeyAndCount =
                             catalog.streams.map { stream ->
                                 val manager = syncManager.getStreamManager(stream.descriptor)
+                                totalRecords = totalRecords.plus(manager.readCount())
```
5. The last step would be to add the byte count to the Checkpoint message on its way *out.* The path for flushing state starts at `CheckpointManager::flushReadyCheckpointMessages`. It branches for Stream-v-Global, but both call `StreamManager::areRecordsPersistedForCheckpoint`. You can probably just use that as a model to make a method that returns the checkpoint value for a given id and add it to the checkpoint before emitting it.

Caveat: `CheckpointManager` is parameterized by `T` (`Any`), which is the message type it emits. So you'll need to make that at least `T: CheckpointMessage`, which will require updating some tests.

subodh1810 committed:
Merge branch 'master' into add-bytes-to-state

# Conflicts:
#	airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/message/DestinationMessage.kt

subodh1810 committed:
address review comments + impl dummy stats emitter

subodh1810 committed:
Merge branch 'master' into add-bytes-to-state

# Conflicts:
#	airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/state/StreamManager.kt

johnny-schmidt commented on airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/message/PipelineEvent.kt:
```diff
@@ -17,7 +17,7 @@ sealed interface PipelineEvent<K : WithStream, T>
  */
 data class PipelineMessage<K : WithStream, T>(
-    val checkpointCounts: Map<CheckpointId, Long>,
+    val checkpointCounts: Map<CheckpointId, Pair<Long, Long>>,
```
This might be more clear as `Map<CheckpointId, CheckpointValue>`

subodh1810 committed:
fix merge conflicts

subodh1810 committed:
fix build

subodh1810 committed:
fix test

subodh1810 committed:
Merge branch 'master' into add-bytes-to-state

johnny-schmidt commented on airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/state/CheckpointManager.kt:
```diff
@@ -169,10 +171,52 @@ class CheckpointManager<T>(
+                            (head.value.checkpointMessage as CheckpointMessage)
+                                .checkpointKey!!
+                                .checkpointId
+                        catalog.streams.map { stream ->
```
this could be simplified to `catalog.streams.map { ... }.sumOf/By { ... }`

subodh1810 committed:
fix test

subodh1810 committed:
does this fix build?

subodh1810 committed:
remove unnecessary casting

johnny-schmidt commented:
This looks good, but the new behavior should be tested. At the very least
* `PipelineEventBookkeepingRouterTest`: adds bytes to the stream manager on record read
* `CheckpointManagerUTest`: adds the stream manager's reported committed counts and byte counts to both stream and global state messages

Other than that, just totally optional nits about the semantics around merge/sumBy versus manually sum.

subodh1810 committed:
fix test

subodh1810 committed:
Merge branch 'master' into add-bytes-to-state

subodh1810 committed:
address review comments

subodh1810 committed:
format

subodh1810 commented:
@johnny-schmidt addressed all the review comments and added tests for the StatsEmitter. The test for validating additionalProperties of stream messages is in `airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/load/write/BasicFunctionalityIntegrationTest.kt`

johnny-schmidt commented on airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/task/internal/LoadPipelineStepTask.kt:
```diff
@@ -135,7 +141,28 @@ class LoadPipelineStepTask<S : AutoCloseable, K1 : WithStream, T, K2 : WithStrea
                                 }
-                                .let { it.copy(inputCount = it.inputCount + 1) }
+                                .let { it.copy(inputRecordCount = it.inputRecordCount + 1) }
+                                .let {
```
This is wrong. 

1. In general, this task is payload-agnostic. It should not be inspecting the input value or making guesses about what it is. It should just be using the data in the envelope. (This is the generic task wrapper for all tasks.)
2. We don't actually need the byte count in `StateWithCount`. That's for tracking *input* counts and data age to perform the test to determine whether we should force a data flush (it's mainly to enforce the time limit, but also to enable microbatching for testing). You can just remove it. You already get what you need in lines 178-180 by adding bytes to the `CheckpointValue`

subodh1810 committed:
inputByteCount is not required in StateWithCounts
